### ISSUE TITLE
Uses outdated .elc files

### ISSUE THREAD

**Issue by alphapapa** [jorgenschaefer/emacs-buttercup#142]
I've noticed that sometimes while developing and testing, my tests may fail because of an outdated `.elc` file that I haven't recompiled yet.  Buttercup warns about this, but the warning quickly scrolls off the screen as it runs the tests.

It would be helpful if there were a simple way to make Buttercup either refuse to run if the byte-compiled file is outdated, or byte-compile it before running the tests.

I could write some code to do this in my test file, but it seems like a common issue, so it would be nice if there were a standard way to solve it.

Thanks.

**Comment by snogge** [comment]
I guess this is the normal Emacs warning about elc-files that are older than the el-file?
Is 

    (setq load-prefer-newer t)

in your test file a good enough solution?

**Comment by alphapapa** [comment]
That does help if you don't mind running interpreted code, but it doesn't help if you want to run byte-compiled code.  It's important to use one or the other, because behavior can differ between them (they are basically two different implementations of Elisp, with subtle differences which occasionally matter).  I can imagine being frustrated if my test suite inconsistently tested `.el` or `.elc` files, causing weird failures which would be hard to track down.

**Comment by jorgenschaefer** [comment]
Isn't this the same thing as for any other language? If you do not recompile before running your tests, you end up running outdated code? I fully understand the frustration, but this seems to be a general Emacs problem, not a particular buttercup problem …

**Comment by alphapapa** [comment]
@jorgenschaefer I could write some custom code to do this in my test files, but it seems like a common issue, so it would be nice if there were a standard way to solve it with the test library.  That's what Buttercup is for--making it easier to test Emacs packages--right?  :)

**Comment by jorgenschaefer** [comment]
Buttercup does not warn about outdated elc files. Emacs does. That's why I am saying that this is an Emacs issue, not a buttercup issue – Emacs (i.e. the Emacs Lisp interpreter) simply does not run the correct Lisp code in these cases. :-(

**Comment by jorgenschaefer** [comment]
Buttercup did not generate the .elc files in the first place, so it's not exactly its place to regenerate them if they're outdated. We could turn those warnings into errors to abort the test run if Emacs would run outdated code, though. (PR welcome :-D)

**Comment by sergeyklay** [comment]
I use the following approach in my tests:

**`test/utils.el`**
```elisp
(require 'buttercup)

(let* ((current-file (if load-in-progress load-file-name (buffer-file-name)))
       (source-directory (locate-dominating-file current-file "Cask"))
       ;; Don't load old byte-compiled versions
       (load-prefer-newer t))
  ;; Load the file under test
  (load (expand-file-name "my-awesome-mode" source-directory)))
```

**`test/my-awesome-mode-test.el`**
```elisp
(require 'utils)
```

**Comment by snogge** [comment]
I tried using a file handler, but for some reason calling the real `load` does not work.

```emacs-lisp
(add-to-list 'file-name-handler-alist '("\\.elc\\'" . buttercup--elc-handler))

(defun buttercup--elc-handler (operation &rest args)
  "Signal an error when loading a too-old elc file.
OPERATION and ARGS according to file-handlers."
  (save-match-data
    (when (eq operation 'load)
      (buttercup--elc-check (car args)))
    (buttercup-run-real-handler operation args)))

(defun buttercup--elc-check (file)
  (let ((elfile (substring file 0 -1)))
    (if (file-exists-p elfile)
        (let ((elc-attributes (file-attributes file))
              (el-attributes (file-attributes elfile)))
          (if (time-less-p (file-attribute-modification-time elc-attributes)
                           (file-attribute-modification-time el-attributes))
              (error "%s is newer than %s" elfile file))))))

;; If we are given an operation that we don't handle,
;; call the Emacs primitive for that operation,
;; and manipulate the inhibit variables
;; to prevent the primitive from calling our handler again.
(defun buttercup--run-real-handler (operation args)
  "Run the real primitive for OPERATION with ARGS."
  (let ((inhibit-file-name-handlers
	     (cons 'buttercup--elc-handler
	           (and (eq inhibit-file-name-operation operation)
		            inhibit-file-name-handlers)))
	    (inhibit-file-name-operation operation))
    (apply operation args)))
```

**Comment by snogge** [comment]
Hi @alphapapa , do you think the fix in #183 would solve this problem?

**Comment by alphapapa** [comment]
@snogge I think that would, yes.  And I think I can apply those hooks in my `makem.sh`, since it calls Buttercup Elisp functions directly.  Thank you for working on that.
